astropy/time/tests/test_precision.py

```diff
@@ -142,69 +293,376 @@ def test_leap_seconds_rounded_correctly():
-        t2 = getattr(t, scale2)
+    try:
+        with quiet_erfa():
+            t2 = getattr(t, scale2)
+    except iers.IERSRangeError:  # UT1 conversion needs IERS data
+        assume(False)
+    except ErfaError:
+        assume(False)
     assert t2.jd1 % 1 == 0
     assert abs(t2.jd2) <= 0.5
     assert abs(t2.jd2) < 0.5 or t2.jd1 % 2 == 0
 
 
-def test_conversion_preserves_jd1_jd2_invariant_2():
-    """Conversion can leave abs(jd2)>0.5"""
-    scale1 = 'tai'
-    scale2 = 'tcb'
-    jd1, jd2 = (2441316.5, 0.0)
+@given(scale1=sampled_from(STANDARD_TIME_SCALES),
+       scale2=sampled_from(STANDARD_TIME_SCALES),
+       jds=unreasonable_jd())
+def test_conversion_never_loses_precision(iers_b, scale1, scale2, jds):
+    jd1, jd2 = jds
     t = Time(jd1, jd2, scale=scale1, format="jd")
-    t2 = getattr(t, scale2)
-    assert t2.jd1 % 1 == 0
-    assert abs(t2.jd2) <= 0.5
-    assert abs(t2.jd2) < 0.5 or t2.jd1 % 2 == 0
+    try:
+        # If the rates ever differ, might lose one ULP
+        with quiet_erfa():
+            t2 = t + 2*dt_tiny
+            assert getattr(t, scale2) < getattr(t2, scale2)
+    except iers.IERSRangeError:  # UT1 conversion needs IERS data
+        assume(scale1 != 'ut1' or 2440000 < jd1 + jd2 < 2458000)
+        assume(scale2 != 'ut1' or 2440000 < jd1 + jd2 < 2458000)
+        raise
+    except ErfaError:
+        # If the generated date is too early to compute a UTC julian date,
+        # and we're not converting between scales which are known to be safe,
+        # tell Hypothesis that this example is invalid and to try another.
+        # See https://docs.astropy.org/en/latest/time/index.html#time-scale
+        barycentric = {scale1, scale2}.issubset({'tcb', 'tdb'})
+        geocentric = {scale1, scale2}.issubset({'tai', 'tt', 'tcg'})
+        assume(jd1 + jd2 >= -31738.5 or geocentric or barycentric)
+        raise
+
+
+@given(sampled_from(leap_second_deltas), floats(0.1, 0.9))
+def test_leap_stretch_mjd(d, f):
+    mjd, delta = d
+    t0 = Time(mjd, format="mjd", scale="utc")
+    th = Time(mjd + f, format="mjd", scale="utc")
+    t1 = Time(mjd + 1, format="mjd", scale="utc")
+    assert_quantity_allclose((t1 - t0).to(u.s), (1 * u.day + delta * u.s))
+    assert_quantity_allclose((th - t0).to(u.s), f * (1 * u.day + delta * u.s))
+    assert_quantity_allclose((t1 - th).to(u.s), (1 - f) * (1 * u.day + delta * u.s))
+
+
+@given(scale=sampled_from(STANDARD_TIME_SCALES),
+       jds=unreasonable_jd(),
+       delta=floats(-10000, 10000))
+@example(scale='utc',
+         jds=(0.0, 2.2204460492503136e-13),
+         delta=6.661338147750941e-13)
+@example(scale='utc',
+         jds=(2441682.5, 2.2204460492503136e-16),
+         delta=7.327471962526035e-12)
+@example(scale='utc', jds=(0.0, 5.787592627370942e-13), delta=0.0)
+def test_jd_add_subtract_round_trip(scale, jds, delta):
+    jd1, jd2 = jds
+    if scale == 'utc' and 1 > abs(jd1+jd2):
+        # near-zero UTC times degrade accuracy; why?
```

Comment:
This is related to the point above.  near-zero UTC leads to having a jd2 near 0.5, and then the conversion to TAI can push it over 0.5 which loses a bit of precision. Not sure why the threshold would not be something like `4*dt_tiny`, though...